<a href="https://colab.research.google.com/github/rosslogan702/hypothesis_testing_notes/blob/master/sample_size_determination.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Hypothesis Testing - Sample Size Determination A/B Testing

# Contents

The focus of this notebook is sample size determination for A/B Testing.

The notebook will cover the following:



1.   Description
2.   Power Analysis & How to Calculate Sample Size Manually
3.   Practical Examples of Sample Size Determination

# 1. Description

Sample size determination for A/B Testing is an important part of any experiment that is planned to be run. 

Running an experiment takes time and effort, and we don't want to run the experiment without having an appropriate sized sample to be able to be confident in the results.

As an example, if we were to ask 10 people who they were voting for in the next US presidential election. The sample size of 10 people would be insufficient to try to predict the results of the election.

There are a number of values that are required to be used as part of a calculation to determine the sample size.

These are:


*   Baseline conversion rate
*   Statistical significance
*   Minimum detectable effect
*   Statistical Signifiance Level
*   Power

An A/B test is a scientific way of choosing between two options.

Some examples of where A/B testing would be used would be:

*  Choosing between two landing pages of a website to determine which one has optimal conversion
*  Choosing between two colors of purchase button on a website for optimal purchases

A simple sample A/B test calculator is shown here: https://s3.amazonaws.com/codecademy-content/courses/learn-hypothesis-testing/a_b_sample_size/index.html


# 2. Power Analysis - How to Calculate the Minimum Sample Size Required Manually





## Description

Power Analysis is an important aspect of experimental design for A/B tests as this allows us to determine the minimum required sample size to detect an effect of a given size at the desired level of significance. 

For the power analysis to be conducted, we need to set some statistical values:

*   statistical significance level ($\alpha$)
*   power level

The statistical significance level is the probability of rejecting the null hypothesis when it is true (no difference exists).

The power level is the probability of rejecting the null hypothesis when it is false.

There is another value called beta ($\beta$) which is the probability of accepting the null hypothesis when it is false. (power = 1- $\beta$)

For the purposes of this notebook we are going to assume what are generally some industry accepted standards, we will set the $\alpha$ level at 0.05 (5%) and the power level at 0.80 (80%), therefore $\beta$ will be 0.20 (20%)

## Calculating Minimum Sample Size Manually

Using a number of online A/B Test Sample Size calculators yields a variation of results in the minimum sample size that is required per variant to be confident in the results.

One such formula for calculating the sample size here but in practical reality many different companies and variations exist for calculating the sample size.

$n= (p_1(1-p_1)) + (p_2(1-p_2)) * (\frac{Z_\text{power} + Z_\text{1-alpha}}{p_1 - p_2})^2$



*   p1 - current conversion rate
*   p2 - current conversion late + desired lift
*   Z is the critical value (or z-score) derived from the standard normal distribution
    *   $Z_\text{power}$
    *  $Z_\text{1-alpha}$
    
 



## Example One

Let's say that we have a current baseline conversion rate of 10% and we want to be able to test a 50% increase in conversion. 

Assuming our signicance level is set at 5% and the power is set to 80%, what would be the minimum sample size required to perform an A/B test?



In [0]:
from scipy.stats import norm
p_1 = 0.10
p_2 = 0.15
p_1_ = 1-p_1
p_2_ = 1-p_2
alpha = 0.05
power_level = 0.80
z_power = norm.ppf(0.8)
z_alpha = norm.ppf(0.95)

In [0]:
n = (p_1*p_1_+p_2*p_2_) * (((z_power + z_alpha)**2)/((p1 - p2)**2))
print('Min Sample Size Required: {0:.3f} per variant'.format(n))

Min Sample Size Required: 537.882


AS the min sample size required, comes out as 537.882, we would need a minimum sample size of 538 per variant.

## Example Two

Let's say that we have a current baseline conversion rate of 20% and we want to be able to test a 30% increase in conversion. 

Assuming our signicance level is set at 5% and the power is set to 80%, what would be the minimum sample size required to perform an A/B test per variant?

In [4]:
from scipy.stats import norm
p_1 = 0.20
p_2 = 0.26
p_1_ = 1-p_1
p_2_ = 1-p_2
alpha = 0.05
power_level = 0.80
z_power = norm.ppf(0.8)
print('z_power: {0:.3f}'.format(z_power))
z_alpha = norm.ppf(0.95)
print('z_alpha: {0:.3f}'.format(z_alpha))

z_power: 0.842
z_alpha: 1.645


In addition to using the norm.ppf function, we can use the following z-table to find the values for 0.80 and 0.95: http://pages.stat.wisc.edu/~ifischer/Statistical_Tables/Z-distribution.pdf

To do this we look for the values of 0.8 and 0.95 in the right column of the table and then find the corresponding z_score on the left column which is what we then put into the formula.

In [3]:
n = (p_1*p_1_+p_2*p_2_) * (((z_power + z_alpha)**2)/((p_1 - p_2)**2))
print('Min Sample Size Required: {0:.3f} per variant'.format(n))

Min Sample Size Required: 605.204 per variant


As the min required sample size variant comes out as 605.204, this means we would need 606 samples per variant.

# 3. Practical Example using Simple Sample Size Calculator

Using the sample size calculator, what would be the sample size required to be able to detect at least a 30% increase from a 20% baseline with 90% statistical significance?

Print the result of the require sample size.

Using the calculator noted above at: https://s3.amazonaws.com/codecademy-content/courses/learn-hypothesis-testing/a_b_sample_size/index.html  

baseline conversion rate - 20%  
minimumum detectable effect - 30%  (this is the uplift figure that we want to detect)
statistical significance - 90%  

Plugging these numbers into the sample size calculator we get a sample size returned of 760